### PR TITLE
Add example plugins for adding Gradle Enterprise and CCUD Gradle plugins

### PR DESCRIPTION
**Pull request by jthurne**

### CODE REVIEW

**Comment by jthurne** [comment]
See the README.md for a description of the project layout.

File: gradle-enterprise-conventions-gradle-plugin/example-builds/gradle_5.6/settings.gradle [hunk 1]
```diff
@@ -0,0 +1,9 @@
+pluginManagement {
+ repositories {
+     mavenLocal()
```
Comment (ldaley) [inline gradle-enterprise-conventions-gradle-plugin/example-builds/gradle_5.6/settings.gradle:3]: To be replaced with project local repo.

mavenLocal() should generally be avoided given that it is unmanaged state.

File: gradle-enterprise-conventions-gradle-plugin/init-script/add-gradle-enterprise-conventions.gradle [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            mavenLocal()
+            gradlePluginPortal()
+        }
+        dependencies {
+            // CHANGE ME: change to match your convention plugin's name
+            classpath("com.myorg:gradle-enterprise-conventions-gradle-plugin:${pluginVersion}")
+        }
+    } else if (majorGradleVersion >= 2) {
+        repositories {
+            mavenLocal()
```
Comment (ldaley) [inline gradle-enterprise-conventions-gradle-plugin/init-script/add-gradle-enterprise-conventions.gradle:29]: Missing a changeme here.

File: gradle-enterprise-conventions-gradle-plugin/plugins/gradle-5-or-newer/src/main/java/com/myorg/GradleEnterpriseConventionsPlugin.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            applySettingsPlugin((Settings) target);
+        } else if (target instanceof Project) {
+            if (isGradle6OrNewer()) {
+                throw new GradleException("For Gradle versions 6.0 and newer, gradle-enterprise-conventions must be applied to Settings");
+            }
+            applyProjectPlugin((Project) target);
+        }
+    }
+
+    private void applySettingsPlugin(Settings settings) {
```
Comment (ldaley) [inline gradle-enterprise-conventions-gradle-plugin/plugins/gradle-5-or-newer/src/main/java/com/myorg/GradleEnterpriseConventionsPlugin.java:32]: Maybe include the Gradle version numbers in these applyMethods to make things a bit clearer.

File: gradle-enterprise-conventions-gradle-plugin/plugins/gradle-5-or-newer/src/main/java/com/myorg/GradleEnterpriseConventionsPlugin.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        gradleEnterprise.setServer("https://ge.myorg.com");
+
+        BuildScanExtension buildScan = gradleEnterprise.getBuildScan();
+        buildScan.publishAlways();
+        buildScan.setCaptureTaskInputFiles(true);
+    }
+
+    private void configureBuildCache(BuildCacheConfiguration buildCache) {
+        // CHANGE ME: Apply your build cache configuration here
+        buildCache.local(local -> {
```
Comment (ldaley) [inline gradle-enterprise-conventions-gradle-plugin/plugins/gradle-5-or-newer/src/main/java/com/myorg/GradleEnterpriseConventionsPlugin.java:60]: I'm not sure we should have any local build configuration here.

File: gradle-enterprise-conventions-gradle-plugin/example-builds/gradle_5.6/settings.gradle (hunk 1 above)
Comment (ldaley) [reply gradle-enterprise-conventions-gradle-plugin/example-builds/gradle_5.6/settings.gradle:3]: Changed me mind. Let's go with this as it's familiar to people.

File: gradle-enterprise-conventions-gradle-plugin/plugins/gradle-2-through-4/src/main/java/com/myorg/BuildScansConventionsPlugin.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+import com.gradle.scan.plugin.BuildScanExtension;
+import com.gradle.scan.plugin.BuildScanPlugin;
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class BuildScansConventionsPlugin implements Plugin<Project> {
```
Comment (ldaley) [inline gradle-enterprise-conventions-gradle-plugin/plugins/gradle-2-through-4/src/main/java/com/myorg/BuildScansConventionsPlugin.java:11]: Maybe worth some explanatory class level Javadoc explaining that this is the plugin used for < Gradle 5 etc.

File: gradle-enterprise-conventions-gradle-plugin/plugins/gradle-5-or-newer/src/main/java/com/myorg/GradleEnterpriseConventionsPlugin.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    private void applyProjectPlugin(Project project) {
+        if (!project.equals(project.getRootProject())) {
+            throw new GradleException("gradle-enterprise-conventions may only be applied to root project");
+        }
+        project.getPluginManager().apply(BuildScanPlugin.class);
+        project.getPluginManager().withPlugin("com.gradle.build-scan", __ -> {
+            configureGradleEnterprise(project.getExtensions().getByType(GradleEnterpriseExtension.class));
+            // Build cache configuration cannot be accessed from a project plugin
```
Comment (ldaley) [inline gradle-enterprise-conventions-gradle-plugin/plugins/gradle-5-or-newer/src/main/java/com/myorg/GradleEnterpriseConventionsPlugin.java:45]: This comment on its own is a bit confusing.

**Review by ldaley — CHANGES_REQUESTED** [review]

File: gradle-enterprise-conventions-gradle-plugin/init-script/add-gradle-enterprise-conventions.gradle (hunk 2 above)
Comment (jthurne) [reply gradle-enterprise-conventions-gradle-plugin/init-script/add-gradle-enterprise-conventions.gradle:31]: Added

File: gradle-enterprise-conventions-gradle-plugin/plugins/gradle-5-or-newer/src/main/java/com/myorg/GradleEnterpriseConventionsPlugin.java (hunk 3 above)
Comment (jthurne) [reply gradle-enterprise-conventions-gradle-plugin/plugins/gradle-5-or-newer/src/main/java/com/myorg/GradleEnterpriseConventionsPlugin.java]: I renamed the methods, and I do think it's a bit clearer now.

File: gradle-enterprise-conventions-gradle-plugin/plugins/gradle-5-or-newer/src/main/java/com/myorg/GradleEnterpriseConventionsPlugin.java (hunk 4 above)
Comment (jthurne) [reply gradle-enterprise-conventions-gradle-plugin/plugins/gradle-5-or-newer/src/main/java/com/myorg/GradleEnterpriseConventionsPlugin.java]: Removed

File: gradle-enterprise-conventions-gradle-plugin/plugins/gradle-2-through-4/src/main/java/com/myorg/BuildScansConventionsPlugin.java (hunk 5 above)
Comment (jthurne) [reply gradle-enterprise-conventions-gradle-plugin/plugins/gradle-2-through-4/src/main/java/com/myorg/BuildScansConventionsPlugin.java:15]: Added

File: gradle-enterprise-conventions-gradle-plugin/plugins/gradle-5-or-newer/src/main/java/com/myorg/GradleEnterpriseConventionsPlugin.java (hunk 6 above)
Comment (jthurne) [reply gradle-enterprise-conventions-gradle-plugin/plugins/gradle-5-or-newer/src/main/java/com/myorg/GradleEnterpriseConventionsPlugin.java]: Comment has been rewritten.

**Pull request merged**